### ISSUE TITLE
How to build from scratch

### ISSUE THREAD

**Issue by kingio** [saiarcot895/chromium-ubuntu-build#44]
Hi!!

The readme looks like a generic one and doesn't tell how to make the build, commands listed like `./debian/rules get-current-source` return `dh: Unknown sequence get-current-source`.

I tried running `debuild -b -uc -us` but it misses the compiled chrome so it clearly didn't compile the build.
Output:
```bash
sed -i 's,/etc/chromium-browser/policies,/etc/chromium/policies,' \
        /chromium-ubuntu-build/chrome/common/chrome_paths.cc
sed: can't read /chromium-ubuntu-build/chrome/common/chrome_paths.cc: No such file or directory
debian/rules:295: recipe for target 'override_dh_clean' failed
```
ls `/chromium-ubuntu-build`:
```bash
debian
ui
```

Other commands like `./debian/rules build` return:
`/bin/bash: tools/gn/bootstrap/bootstrap.py: No such file or directory`

Could you write the instructions?

**Comment by omen23** [comment]
afaik .. this is just the ubuntu build layer (debian packaging etc) but you need to fetch the chromium source yourself then you can try `debuild --no-sign -b` ...
Greets

**Comment by saiarcot895** [comment]
Hmm, I should probably update the README.source at some point. Here's a summarized list of steps:

1. Run `ORIG_VERSION=73.0.3683.10 debian/rules get-packaged-orig-source`, where `ORIG_SOURCE` is the version you want to get.
2. Extract the tarball.
3. Run `quilt push -a` to apply all patches.
4. Run `debuild` or `dpkg-package` as you normally would.

**Comment by timrichardson** [comment]
Please do write them up.
When I try the steps above, using exactly the same version:
"Extract the tarball". This makes a new subdirectory; I move the debian directory into it. 

I try
`debuild -b -uc -us
`
which builds the source package, but it ends with 

LINK gn
FAILED: gn 
/usr/bin/clang++ -Wl,-Bsymbolic-functions -Wl,-z,relro -O3 -fdata-sections -ffunction-sections -Wl,--gc-sections -Wl,-strip-all -Wl,--icf=all -static-libstdc++ -Wl,--as-needed -o gn -Wl,--start-group tools/gn/gn_main.o base.a gn_lib.a -Wl,--end-group -ldl -lpthread
/usr/bin/ld: unrecognised option '--icf=all'

When I download the source package from the ppa, after the package downloads, it seems to apply a  patch which says "dpkg-source: info: applying gn-remove-icf.patch"

but it seems that I am not getting this patch applied in step 3 when using the chromium-ubuntu-build package. Am I doing something wrong or is this a real issue?

**Comment by omen23** [comment]
idk does this build with clang?

Am Sa., 23. Feb. 2019 um 12:04 Uhr schrieb Tim Richardson <
notifications@github.com>:

**Comment by saiarcot895** [comment]
@timrichardson After you moved the debian directory into the subdirectory (you can do it the other way around as well; i.e. move the contents of the subdirectory to the current directory), did you run `quilt push all`?

@omen23 Yes, it uses the system clang binary. Compiling with GCC might work if it's new enough. With the current code, actually, it only successfully compiles if I use their version of clang.

**Comment by saiarcot895** [comment]
@timrichardson Never mind, I just updated the git repository with my changes. Can you try now?

**Comment by timrichardson** [comment]
Now
quilt push -a
No series file found 
(I moved the contents of the extracted tar to its parent directory, so I did not move the debian folder)

set export QUILT_PATCHES=debian/patches
makes no difference

**Comment by timrichardson** [comment]
I followed these instructions to setup .quiltrc and it quilt push -a worked

https://raphaelhertzog.com/2012/08/08/how-to-use-quilt-to-manage-patches-in-debian-packages/

**Comment by timrichardson** [comment]
build of ORIG_VERSION=73.0.3673.0 (which I used because I saw it reference in a recent commit to this repository)
eventually fails with:

ninja: error: '../../chrome/test/data/webui/web_ui_test.mojom', needed by 'gen/chrome/test/data/webui/web_ui_test.mojom-shared-message-ids.h', missing and no known rule to make it
debian/rules:218: recipe for target 'override_dh_auto_build' failed

**Comment by saiarcot895** [comment]
The patches in the latest version are based on 73.0.3683.10. Please use that version for `ORIG_VERSION`.

The error you're hitting was an issue in Chromium's generation of the source tarballs (i.e. not a build error on your part). That was fixed in a later version (I think 73.0.3679.0).

**Comment by timrichardson** [comment]
Thanks. Now all patches applied cleanly. It moves into the large compile phase with 
Next problem: 
/bin/sh: 1: /usr/bin/clang-4.0: not found
my default clang is v6, and the deb source compiles with it. Is this a problem with configuration?

[7/21107] CC obj/base/third_party/libevent/libevent/evrpc.o
FAILED: obj/base/third_party/libevent/libevent/evrpc.o 
/usr/bin/clang-4.0 -MMD -MF obj/base/third_party/libevent/libevent/evrpc.o.d -DHAVE_CONFIG_H -DUSE_UDEV -DUSE_AURA=1 -DUSE_GLIB=1 -DUSE_NSS_CERTS=1 -DUSE_X11=1 -DNO_TCMALLOC -DFULL_SAFE_BROWSING -DSAFE_BROWSING_CSD -DSAFE_BROWSING_DB_LOCAL -DCHROMIUM_BUILD -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_GNU_SOURCE -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -I../../base/third_party/libevent/linux -I../.. -Igen -fno-strict-aliasing --param=ssp-buffer-size=4 -fstack-protector -Wno-builtin-macro-redefined -D__DATE__= -D__TIME__= -D__TIMESTAMP__= -funwind-tables -fPIC -pthread -fcolor-diagnostics -fmerge-all-constants -no-canonical-prefixes -m64 -march=x86-64 -O2 -fno-ident -fdata-sections -ffunction-sections -fno-omit-frame-pointer -g2 -ggnu-pubnames -fvisibility=hidden -Wheader-hygiene -Wstring-conversion -Wtautological-overlap-compare -Wall -Wno-unused-variable -Wno-missing-field-initializers -Wno-unused-parameter -Wno-c++11-narrowing -Wno-unneeded-internal-declaration -Wno-undefined-var-template -std=c11 -c ../../base/third_party/libevent/evrpc.c -o obj/base/third_party/libevent/libevent/evrpc.o
/bin/sh: 1: /usr/bin/clang-4.0: not found

**Comment by timrichardson** [comment]
The deb source .gn/applied-patches shows this patch applied:
use-clang-versioned-bionic.patch

the git repo .gn/applied-patches shows this patch applied:
use-clang-versioned.patch

**Comment by saiarcot895** [comment]
I have two separate series files in use, depending on whether it's building for Ubuntu Xenial or for Ubuntu Bionic or newer. The reason is that on Xenial, the default version of clang isn't good enough to build Chromium. Since you're on Bionic (or the equivalent or newer), there's two ways to fix this. The quick and dirty way is to edit `build/toolchain/gcc_toolchain.gni`, look for `clang-4.0` and `clang++-4.0` near the end of the file, and replace those with `clang` and `clang++` respectively. However, you won't be able to cleanly unapply all of the patches afterwards.

The proper way is to follow these steps:

1. Run `quilt pop -a` to unapply all patches.
2. `mv debian/patches/series.bionic debian/patches/series`
3. `quilt push -a`

And then restart the build.

**Comment by timrichardson** [comment]
ah ok. I'm learning a lot. Meanwhile, it built with clang 4. But it seems that you have to pick the chrome releases carefully. 
I will write up everything I've learnt and send a patch, it will be Building Chromium for Dummies.

**Comment by timrichardson** [comment]
PS I see with v73 

chrome://media-internals/
reports
video_decoder | MojoVideoDecoder

It's definitely doing hardware decoding: with a 4K test video I'm doing about 7% CPU.  
-- | --

**Comment by timrichardson** [comment]
I built 73.0.3683.39 successfully, using the bionic patch series.
 The gn-remove-icf.patch isn't needed any more (or at the moment).
These are my notes 'for dummies' https://docs.google.com/document/d/1EtRbk13YugJyi0oAHtYDEDRQ05kyb8H2WyasqOdbHdc/edit?usp=sharing
note that 
tar -xf chromium-browser_73.0.3683.39.orig.tar.xz --strip 1
is the best way I found to extract the tar; it skips the containing subdirectory.
